File: language-support/js/daml-ledger/src/validation/variant.ts
// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

import * as ledger from "..";
import { native, object, Validation } from "./base";
import { Identifier } from "./identifier";
import { Value } from './value'
import { RequiredValidation, OptionalValidation } from "./base/typelevel";

function required(): RequiredValidation<ledger.Variant> {
    return {
        constructor: native('string'),
        value: Value,
    };
}

function optional(): OptionalValidation<ledger.Variant> {
    return {
        variantId: Identifier,
    };
}

export const Variant: Validation = object('Variant', required, optional);

File: sdk/language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/Variant.java
// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

package com.daml.ledger.javaapi.data.codegen;

/**
 * Base class of all decoded-to-codegen Daml variants <em>with no type parameters</em>.
 *
 * <p>Its encoded counterpart is {@link com.daml.ledger.javaapi.data.Variant}, which can be produced
 * with {@link #toValue}.
 *
 * @param <T> A "self type", some subclass of this class that {@code T} implements.
 */
public abstract class Variant<T> implements DefinedDataType<T> {
  public abstract com.daml.ledger.javaapi.data.Variant toValue();
}
